42 score mechanics

joshjwilli opened the pull request:
- Fully implemented the issue as described
- All score displays are placeholder and not intended to be final
- Placeholder function for when bosses are added
- ChangeScore should enable easy addition of any other point modifiers (i.e. the card system)
- Added a placeholder infinite level and got it working in-game for testing and debugging purposes of the new score mechanics

joshjwilli committed:
Per Level Score Tracking Implemented

- Both levels should be able to interact with their own score manager script to properly record the points gained from felling enemies or having positive distance in the level
- Each level currently saves it's highscore to player prefs
- Still need to implement overall score and displaying of the scores

joshjwilli committed:
Implemented score display in level and in level select

Not intended to be final UI for score display, but a good placeholder and good for testing functionality and balancing

joshjwilli committed:
Fully Implemented Score Mechanics and Other Placeholders

- Fully implemented as described in the issue
- Placeholder function for when bosses are added
- ChangeScore(float pts) should enable quick addition of any other score modifiers (i.e. when the card system is flushed out)
- Also added a placeholder infinite level and got functionality on the main menu for the purposes of ensuring score works across the board

joshjwilli committed:
Deleted unused script

joshjwilli committed:
Added Comment for clarity

adenprince commented on Assets/Scripts/Score/OverallScoreManager.cs:
```diff
@@ -0,0 +1,21 @@
+    private TextMeshProUGUI highScore;
+
+    private float overallScore;
+    private int lastLevel = 2;
```
Currently, this variable duplicates the `lastLevel` constant in `LevelEndTriggerController.cs`. I think it's better to add `Constants.cs` and use a public constant there in both this script and `LevelEndTriggerController.cs`.

joshjwilli replied:
Agreed

adenprince commented on Assets/Scenes/Levels/Level1.unity:
Score text should be anchored to the top of the screen (like the pause button) so that it's location relative to the stop of the screen doesn't change with the screen height.

adenprince commented on Assets/Scenes/Levels/Level1.unity:
Score text should be anchored to the top of the screen (like the pause button) so that its location relative to the top of the screen doesn't change with the screen height.

adenprince commented on Assets/Scripts/Game Controllers/LevelEndTriggerController.cs:
```diff
@@ -10,13 +10,22 @@ public class LevelEndTriggerController : MonoBehaviour
     {
         if (collision.CompareTag("Player"))
         {
+            ScoreManager.Instance.LevelEnd(collision.transform.position.y);
```
I think it's better for `ScoreManager` to have a reference to the player's transform. Then, `ScoreManager` can get the start and end Y positions of the player rather than that being handled outside of that script.

adenprince commented on Assets/Scripts/Score/ScoreManager.cs:
```diff
@@ -0,0 +1,71 @@
+                  displayScore,
+                  pastScore,
+                  startY,
+                  transitionSpeed = 100.0f;
```
`transitionSpeed` -> `displayScoreTransitionSpeed`

adenprince commented on Assets/Scripts/Score/ScoreManager.cs:
```diff
@@ -0,0 +1,71 @@
+    public static ScoreManager Instance { get; private set; }
+    public float Score
+    {
+        get { return score; }
```
`get { return score; }` -> `get => score;`

adenprince commented on Assets/Scripts/Score/ScoreManager.cs:
```diff
@@ -0,0 +1,71 @@
+    private TextMeshProUGUI scoreText;
+
+    private const float moveMulti4Pts = 10.0f, // The value to multiply the distance traveled by
+                        enemyFelledPts = 100.0f; // The points killing an enemy is worth
```
I think that these values should be set in the Inspector (probably also `transitionSpeed`). Then, there should be a Score Manager prefab with an instance in each level so that these values are set in one place.

adenprince replied:
It's good for configuration values to be settable in the Inspector. Then, code familiarity is not needed for changing those values or determining what can safely be changed. In this case, it doesn't matter that much.

adenprince commented on Assets/Scripts/Score/ScoreManager.cs:
```diff
@@ -0,0 +1,71 @@
+    [SerializeField]
+    private TextMeshProUGUI scoreText;
+
+    private const float moveMulti4Pts = 10.0f, // The value to multiply the distance traveled by
```
I think this name could be clearer. Maybe `playerDistancePtsMultiplier`.

adenprince commented on Assets/Scripts/Score/ScoreManager.cs:
```diff
@@ -0,0 +1,71 @@
+    private TextMeshProUGUI scoreText;
+
+    private const float moveMulti4Pts = 10.0f, // The value to multiply the distance traveled by
+                        enemyFelledPts = 100.0f; // The points killing an enemy is worth
```
This comment is unnecessary. (I think the comment about `moveMulti4Pts` could also be removed after that variable is renamed.)

adenprince commented on Assets/Scripts/Score/ScoreManager.cs:
```diff
@@ -0,0 +1,71 @@
+
+    private float score,
+                  displayScore,
+                  pastScore,
```
`pastHighScore` is a little clearer because only the high scores persist.

adenprince commented on Assets/Scripts/Score/ScoreManager.cs:
```diff
@@ -0,0 +1,71 @@
+        ChangeScore(pts);
+    }
+
+    public void AbruptLevelEnd()
```
This method can be removed once `ScoreManager` has a reference to the player's transform that is set in the Inspector.

adenprince commented on Assets/Scripts/Game Controllers/LevelEndTriggerController.cs:
```diff
@@ -10,13 +10,22 @@ public class LevelEndTriggerController : MonoBehaviour
-        indexForNextScene = SceneManager.GetActiveScene().buildIndex + 1;
+        if (SceneManager.GetActiveScene().buildIndex == 0)
+        {
+            indexForNextScene = lastLevel + 1;
```
I don't think this logic is needed in the long term because an infinite level wouldn't have a Level End Trigger. I think there should at least be a comment with that information here. Making the infinite level end without a Level End Trigger is getting into infinite level implementation, so I don't think it needs to be changed in this pull request.

adenprince commented on Assets/Scripts/Score/OverallScoreManager.cs:
```diff
@@ -0,0 +1,21 @@
+
+    private void Awake()
+    {
+        for (int i = 0; i <= lastLevel; i++)
```
I would prefer for the infinite level score to be stored with a PlayerPrefs key that is clearer than `lvl0HighScore`, but changing that does complicate the implementation. I'm fine with keeping this implementation. `lvl0HighScore` being used for the infinite level score should be documented with a comment here.

joshjwilli replied:
Unsure how I would do that without making it's own script, also unsure if that would be worth the extra script.

joshjwilli replied:
Realized can do this through Serializable field, mbad

adenprince commented on Assets/Scripts/Score/ScoreDisplayManager.cs:
```diff
@@ -0,0 +1,22 @@
+    [SerializeField]
+    private TextMeshProUGUI highScore,
+                            buttonTxt;
+    private char level;
```
I think level should just be an int that is set in the Inspector rather than being calculated from the button text. The current implementation relies on level buttons having text that is set in a certain way (even if that button format isn't likely to change).

adenprince commented on Assets/Scripts/Score/ScoreManager.cs:
```diff
@@ -0,0 +1,71 @@
+    private float score,
+                  displayScore,
+                  pastScore,
+                  startY,
```
`startY` -> `startPlayerYPos` or `startPlayerYPosition` (or something like that).

adenprince commented on Assets/Scripts/Score/ScoreManager.cs:
```diff
@@ -0,0 +1,71 @@
+        if (curY > startY) { ChangeScore(moveMulti4Pts * (curY - startY)); }
+
+        Score = (pastScore > Score) ? pastScore : Score;
+        PlayerPrefs.SetFloat("lvl" + SceneManager.GetActiveScene().buildIndex + "HighScore", Score);
```
You could set `"lvl" + SceneManager.GetActiveScene().buildIndex + "HighScore"` to an instance variable in `Awake` to avoid duplicating that code here . I don't think it matters much.

adenprince commented on Assets/Scripts/Score/OverallScoreManager.cs:
```diff
@@ -0,0 +1,21 @@
+    [SerializeField]
+    private TextMeshProUGUI highScore;
+
+    private float overallScore;
```
`overallScore` can be a local variable in `Awake`.

joshjwilli committed:
Implemented half of the requested changes

Shortened a few of the requested variable name changes, can change them to the full version if wanted.

joshjwilli committed:
Finished other half of requested changes

adenprince commented on Assets/Scenes/LevelSelect.unity:
`level` needs to be set for the level 10 button Score Display Manager.

adenprince commented on Assets/Scripts/Score/ScoreManager.cs:
```diff
@@ -0,0 +1,78 @@
+public class ScoreManager : MonoBehaviour
+{
+    public static ScoreManager Instance { get; private set; }
+    public float Score
```
I think this property should be removed. It's unnecessary at this point.

adenprince commented on Assets/Scripts/Score/ScoreManager.cs:
```diff
@@ -0,0 +1,78 @@
+    private string sceneHighScore;
+    
+    // Pass a positive for increase, negative for decrease
+    public void ChangeScore(float amount) { Score += amount; }
```
I think this method should be removed. Instead, the score can be added to directly.

joshjwilli replied:
Removing this would lead to no public option to change the private score variable, do we not want that functionaility?

adenprince replied:
Public access to changing the score isn't needed at this point, but I'm fine with keeping the method. I'm not sure if this method will be used outside of this script in the future. You could also rename `ChangeScore` to something like `AddToScore` and remove the comment.

adenprince commented on Assets/Scripts/Score/ScoreManager.cs:
```diff
@@ -0,0 +1,78 @@
+                  pastHighScore,
+                  startPlayerYPos;
+
+    private string sceneHighScore;
```
`sceneHighScore` -> `sceneHighScoreKey`

adenprince commented on Assets/Scripts/Score/ScoreManager.cs:
```diff
@@ -0,0 +1,78 @@
+            ChangeScore(playerDistPtsMultiplier * (playerTransform.position.y - startPlayerYPos));
+        }
+
+        Score = (pastHighScore > Score) ? pastHighScore : Score;
```
`pastHighScore` can be a local variable here.

adenprince commented on Assets/Scripts/Score/ScoreManager.cs:
```diff
@@ -0,0 +1,78 @@
+        }
+
+        Score = (pastHighScore > Score) ? pastHighScore : Score;
+        PlayerPrefs.SetFloat(sceneHighScore, Score);
```
This is a redundant operation when `Score` is less than `pastHighScore`. I think this should be something like:
```csharp
if (Score > pastHighScore)
{
    PlayerPrefs.SetFloat(sceneHighScore, Score);
}
```

adenprince replied:
I think it would be good to round the score here before saving it (and store the score as an integer using PlayerPrefs). Otherwise, the total displayed score may not be equal to the sum of the displayed level scores because all of the displayed scores are whole numbers.

adenprince commented on Assets/Scenes/LevelSelect.unity:
(There should be a Level Button prefab. Changing that is kind of unrelated to this pull request.)

joshjwilli replied:
Should I implement this, or should it be it's own issue? Maybe an issue of just making pre-fabs for stuff like this?

adenprince replied:
You don't have to implement this. I'll just make this change on `premain`.

adenprince commented on Assets/Scenes/Levels/Level1.unity:
(There should be a Game Controllers prefab. Changing that is kind of unrelated to this pull request.)

joshjwilli replied:
Should I implement this, or should it be it's own issue? Maybe an issue of just making pre-fabs for stuff like this?

adenprince replied:
You don't have to implement this. I'll just make this change on `premain`.

adenprince reviewed (changes requested):
Thank you for making the requested changes. I made a few comments about additional changes.

joshjwilli committed:
Implemented most changes

Waiting on a few responses for the rest

adenprince commented on Assets/Scripts/Score/ScoreManager.cs:
```diff
@@ -44,35 +38,35 @@ public void UpdateScoreDisplay()
         if (buildIndex == 0)
         {
-            sceneHighScore = "lvlIHighScore";
+            sceneHighScoreKey = "lvlIHighScore";
```
I think this key should be changed to `lvlInfHighScore` or `levelInfiniteHighScore` since `lvlI` is less clear and could be mistaken for level 1.

joshjwilli committed:
Finished changes

-Decided to remove ChangeScore()

adenprince commented:
Looks good. Thank you.
